docs/design-notes/raft-in-scylla.md

```diff
@@ -0,0 +1,207 @@
+of schema changes. The group 0 Raft log is used to share all schema
+changes information across all cluster nodes, and perform schema changes
+in the same order on every node. However, while Raft guarantees
+that if an operation succeeds, it's stored in the logs on the
+majority of nodes, it doesn't provide an API for strictly ordered
+schema changes out of the box, for two reasons:
+    - Raft only stores a log of the operations. The operations
+      themselves - the schema changes - are applied to the client
+      state machine once they are committed to the log. In order
+      to construct a new operation it's necessary to read the
+      latest state of the client state machine, and Raft itself
+      doesn't guarantee that the state of the local
+      client state  machine is the *latest*.  E.g. if a client
+      wants to create a table, it's necessary to learn first that
+      the table hasn't been created already. Two clients could read
+      their local state machines, decide that the table does not
+      exist yet, and append identical commands for creating the
+      table to the Raft log. The second command will fail to apply
+      to the client state machine.
+    - if a leader or network fails when committing an operation to
+      Raft log, the client has no way of knowing its status. E.g. a
+      network can time out, and establishing whether or not the
+      majority of Raft nodes store the command in its log or have applied it
+      may be difficult.
+
+For the two reasons above, Scylla uses an additional
+algorithm for all schema changes which it propagates through Raft
+which provides an ordered, at-most-once semantics of command
+application.
+
+This algorithm is using an own system table - `group0_history` - to
+store its state.
+
+The algorithm is adding a unique, monotonic, persistent state id
+to each command that is committed to Raft log of the group 0 state
+machine. This id is used
+to ensure that no two concurrent commands based on the same (and
+outdated) state of the client state machine are applied, and to
+verify that a command is committed to Raft log in a situation of
+uncertainty and retry the command if it failed.
+
+Here's how it works.
+
+Each change to the schema state machine is "signed", or augmented, with a
+state id - a unique monotonic identifier of the change. The state
+id is based on TIMEUUID. When the command is applied, the state id is
+persisted in `group0_history` table. In addition to the new state id,
+the previous state id is also saved in the command to check that
+the command is applied against the correct state of the client
+state machine.
+
+In order to perform a schema change, the client, which can be any Scylla
+node, performs the following steps:
+
+1. Takes a lock around its local client state machine, to protect against
+   concurrent operations of the state machine by different threads of
+   this node.
+1. Issues a raft read barrier. This ensures this local client
```

Comment:
Is the numbering broken or do we assume that the markdown is somehow rendered later?

Reply:
yes